src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependencyTreeTelemetryService.cs

```diff
@@ -0,0 +1,155 @@
+
+        private readonly UnconfiguredProject _project;
+        private readonly ITelemetryService _telemetryService;
+        private readonly ConcurrentDictionary<ITargetFramework, TelemetryState> _telemetryStates =
+            new ConcurrentDictionary<ITargetFramework, TelemetryState>();
+        private readonly object _stateUpdateLock = new object();
+        private string _projectId;
+        private bool _stopTelemetry = false;
+
+        [ImportingConstructor]
+        public DependencyTreeTelemetryService(
+            UnconfiguredProject project,
+            ITelemetryService telemetryService)
+        {
+            _project = project;
+            _telemetryService = telemetryService;
+        }
+
+        /// <summary>
+        /// Indicate whether we have seen all rules we initialized with, in all target frameworks
+        /// </summary>
+        public bool ObservedAllRules() => _telemetryStates.All(state => state.Value.ObservedAllRules());
+
+        /// <summary>
+        /// Initialize telemetry state with the set of rules we expect to observe for target framework
+        /// </summary>
+        public void InitializeTargetFrameworkRules(ITargetFramework targetFramework, IEnumerable<string> rules)
+        {
+            lock (_stateUpdateLock)
+            {
+                if (_stopTelemetry) return;
+
+                var telemetryState = _telemetryStates.GetOrAdd(targetFramework, (key) => new TelemetryState());
+
+                foreach (var rule in rules)
+                {
+                    telemetryState.InitializeRule(rule);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Indicate that a set of rules has been observed in either an Evaluation or Design Time pass.
+        /// This information is used when firing tree update telemetry events to indicate whether all rules
+        /// have been observed.
+        /// </summary>
+        public void ObserveTargetFrameworkRules(ITargetFramework targetFramework, IEnumerable<string> rules)
+        {
+            lock (_stateUpdateLock)
+            {
+                if (_stopTelemetry) return;
+
+                if (_telemetryStates.TryGetValue(targetFramework, out var telemetryState))
+                {
+                    foreach (var rule in rules)
+                    {
+                        telemetryState.ObserveRule(rule);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Fire telemetry when dependency tree completes an update
+        /// </summary>
+        /// <param name="hasUnresolvedDependency">indicates if the snapshot used for the update had any unresolved dependencies</param>
+        public void ObserveTreeUpdateCompleted(bool hasUnresolvedDependency)
+        {
+            bool observedAllRules;
+            lock (_stateUpdateLock)
+            {
+                if (_stopTelemetry) return;
+                _stopTelemetry = !hasUnresolvedDependency;
```

Comment:
>hasUnresolvedDependency [](start = 34, length = 23)

How you stop telemetry if project has unresolved dependency?

Reply:
Should we combine it with observedAllRules flag?

---
In reply to: [150085491](https://github.com/dotnet/project-system/pull/2944#discussion_r150085491) [](ancestors = 150085491)